jobserver/views/users.py

```diff
@@ -1,9 +1,23 @@
 from django.contrib import messages
 from django.contrib.auth.decorators import login_required
+from django.core.exceptions import SuspiciousOperation
+from django.shortcuts import render
 from django.utils.decorators import method_decorator
-from django.views.generic import UpdateView
+from django.views.defaults import bad_request
+from django.views.generic import TemplateView, UpdateView
 
 from ..forms import SettingsForm
+from ..utils import is_safe_path
+
+
+class Login(TemplateView):
```

Comment:
Just a comment for info, rather than anything that needs changing, but you're not really using the TemplateView class here at all so this could just be a plain view function like this:
```py
def login_view(request):
    if is_safe_path(request.GET.get("next", "")):
        return render(request, "login.html")
    else:
        return bad_request(request, SuspiciousOperation)
```
Then in `urls.py` and elsewhere you'd just reference it as `login_view` rather than `LoginView.as_view()`.

Class-based views offer some very powerful features, but there's a conceptual simplicity to function-based views which might help you on your Django journey -- they're just functions which accept a `request` object and return a `response`.

Reply:
Just make sure you return a `TemplateResponse` (it's the same API) rather than a bare `render`! (it adds template and context info to the response, which makes it easier to test)

Reply:
Oh good shout, thanks George.

Reply:
Okay, so `render` returns a `HttpResponse` but this is less convenient than a `TemplateResponse`, because the latter is easier to test. From [Django's documentation](https://docs.djangoproject.com/en/3.2/ref/template-response/), this is because...

> Unlike basic `HttpResponse` objects, `TemplateResponse` objects retain the details of the template and context that was provided by the view to compute the response. The final output of the response is not computed until it is needed, later in the response process.

Unless you feel very strongly about view functions vs class-based views, I'm going to retain the class-based view. Thank you for the comment, though 👍🏻

Reply:
To add to Dave's point, you're wiping out what TemplateView does for you by overriding all of `get()` ([full code here](http://ccbv.co.uk/projects/Django/3.1/django.views.generic.base/TemplateView/#get)).  However, since you need to add `next_url` to the context it all gets a bit bogged down with boilerplate if you try to keep the bits TemplateView is providing for you (context generation, template rendering, and response creation).

If you want to stick with a CBV then View is your best bet:

```
class Login(View):
    def get(self, request, *args, **kwargs):
        next_url = request.GET.get("next", "/")
        if is_safe_path(next_url):
            return TemplateResponse(request, self.template_name, {"next_url": next_url})
        else:
            return bad_request(request, SuspiciousOperation)
```

Reply:
I've converted the CBV to a view function. I'll also read up on Django views. Thanks for the comments @evansd and @ghickman.